src/main/scala/mesosphere/marathon/state/GroupManager.scala

```diff
@@ -202,10 +202,11 @@ class GroupManager @Inject() (
       }
   }
 
-  //scalastyle:off method.length
+  //scalastyle:off method.length cyclomatic.complexity
   private[state] def assignDynamicServicePorts(from: Group, to: Group): Group = {
     val portRange = Range(config.localPortMin(), config.localPortMax())
-    var taken = from.transitiveApps.flatMap(_.portNumbers) ++ to.transitiveApps.flatMap(_.portNumbers)
+    var taken = (from.transitiveApps.flatMap(app => app.hostPorts.flatten ++ app.servicePorts) ++
+      to.transitiveApps.flatMap(app => app.hostPorts.flatten ++ app.servicePorts)).toSet
```

Comment:
TODO: write unit test for this change.
Symptom: multiple apps using port mappings that didn't specify service ports would result in Marathon attempting to reassign the same service port (that was already assigned to app1) to app2. (e.g. `portMappings: [{}]`)